### PR TITLE
Fix mempool DoS vulnerability from malleated transactions

### LINKED ISSUE #8279: Mempool DoS risk in segwit due to malleated transactions
Noticed this in my code review: https://petertodd.org/2016/segwit-consensus-critical-code-review#peer-to-peer-networking

Basically it looks like an attacker may be able to send nodes transactions with malleated witness data, which we don't consistently mark as possibly corrupted in AcceptToMemoryPool(). Result would be those txids being added to recentRejects, messing up propagation.

That said, I haven't actually tested this yet; about to go off to a conference, so if someone else wants to confirm for me that'd be much appreciated! Or if I'm wrong you're welcome to all laugh at my expense. :)

CC: @sdaftuar @sipa

### PR DESCRIPTION
**Pull request by sdaftuar**
~~Fixes #8279~~

In addition to the problem highlighted in that issue, there's an additional, related problem in the sigops policy check.  Because witness sigops are counted without checking that the witness program matches the commitment in the scriptPubKey being spent, it's possible to change a transaction's witness to cause the sigops policy check to fail, without changing the txid.

Similarly, because the bytes-per-sigop check is affected by the size of the transaction including the witness, it's possible to even remove a witness and cause that sigops check to fail, again without changing the txid.

~~So this PR does the following:~~
- ~~Moves the IsStandard check to happen after checking for premature-witness.  (This should prevent the bug reported in #8279 from possibly affecting 0.13.0 nodes, which should never accept witness transactions.)~~
- ~~Changes IsStandard to set a bool which will indicate if the transaction could be malleated, so that the caller can act appropriately.~~
- ~~Reorders the checks in IsStandard so that the size check is performed last, and sets the could-be-malleated bool if the test fails.~~
- ~~Changes the error for sigops failure to always set the could-be-malleated flag.~~
- ~~Adds tests to p2p-segwit.py to catch both scenarios.~~

### CODE REVIEW

**Comment by sdaftuar** [comment]
After more investigation, I've concluded that fixing this for segwit is more complicated than the couple of patches here (I will document the issues more fully in #8279).  However, to fix this issue for 0.13.0, we can simply move the `IsStandard()` check to happen after the premature-witness check.  I will update this PR shortly.

We can separately consider the best way to fix these types of issues more generally after branching off for 0.13.

**Comment by sdaftuar** [comment]
Updated with a simple fix for 0.13.0.

**Comment by laanwj** [comment]
Thanks for fixing this problem and adding a test, too
utACK 46c9620

File: qa/rpc-tests/p2p-segwit.py [hunk 1]
```diff
@@ -68,7 +69,7 @@ def on_pong(self, conn, message):
 
     def on_reject(self, conn, message):
         self.last_reject = message
-        #print message
+        #print (message)
```
Comment (petertodd) [inline qa/rpc-tests/p2p-segwit.py]: Should be `print(message)`

**Comment by petertodd** [comment]
Lightly tested ACK https://github.com/bitcoin/bitcoin/pull/8312/commits/46c9620f11acfd2b528959d6cbab324105c3adef

File: qa/rpc-tests/p2p-segwit.py (hunk 1 above)
Comment (sipa) [inline qa/rpc-tests/p2p-segwit.py]: Or just remove.

**Comment by sipa** [comment]
utACK 46c9620f11acfd2b528959d6cbab324105c3adef

File: qa/rpc-tests/p2p-segwit.py (hunk 1 above)
Comment (MarcoFalke) [inline qa/rpc-tests/p2p-segwit.py]: It is just a comment. Don't consider it a blocker of anything.

**Pull request merged**